lib/HALSim/IMU.cpp

```diff
@@ -0,0 +1,187 @@
+}
+
+void IMU::readAccelerometer()
+{
+    if (nullptr != m_accelerometer)
+    {
+        const double* accelerationValues = m_accelerometer->getValues();
+        if (nullptr != accelerationValues)
+        {
+            m_accelerationValues.valueX = convertFromDoubleToInt16(accelerationValues[0]); /** Index 0 is the x-axis. */
+            m_accelerationValues.valueY = convertFromDoubleToInt16(accelerationValues[1]); /** Index 1 is the y-axis. */
+            m_accelerationValues.valueZ = convertFromDoubleToInt16(accelerationValues[2]); /** Index 2 is the z-axis. */
+        }
+    }
+}
+
+void IMU::readGyro()
+{
+    if (nullptr != m_gyro)
+    {
+        const double* gyroValues = m_gyro->getValues();
+
+        if (nullptr != gyroValues)
+        {
+            m_gyroValues.valueX = convertFromDoubleToInt16(gyroValues[0]); /** Index 0 is the x-axis. */
+            m_gyroValues.valueY = convertFromDoubleToInt16(gyroValues[1]); /** Index 1 is the y-axis. */
+            m_gyroValues.valueZ = convertFromDoubleToInt16(gyroValues[2]); /** Index 2 is the z-axis. */
+        }
+    }
+}
+
+void IMU::readMagnetometer()
+{
+    if (nullptr != m_magnetometer)
+    {
+        const double* magnetometerValues = m_magnetometer->getValues();
+
+        if (nullptr != magnetometerValues)
+        {
+            m_magnetometerValues.valueX = convertFromDoubleToInt16(magnetometerValues[0]); /** Index 0 is the x-axis. */
+            m_magnetometerValues.valueY = convertFromDoubleToInt16(magnetometerValues[1]); /** Index 1 is the y-axis. */
+            m_magnetometerValues.valueZ = convertFromDoubleToInt16(magnetometerValues[2]); /** Index 2 is the z-axis. */
+        }
+    }
+}
+
+void IMU::getAccelerationValues(IMUData* accelerationValues)
+{
+    if (nullptr != accelerationValues)
+    {
+        accelerationValues->valueX = m_accelerationValues.valueX;
+        accelerationValues->valueY = m_accelerationValues.valueY;
+        accelerationValues->valueZ = m_accelerationValues.valueZ;
+    }
+}
+
+void IMU::getTurnRates(IMUData* turnRates)
+{
+    if (nullptr != turnRates)
+    {
+        turnRates->valueX = m_gyroValues.valueX;
+        turnRates->valueY = m_gyroValues.valueY;
+        turnRates->valueZ = m_gyroValues.valueZ;
+    }
+}
+
+void IMU::getMagnetometerValues(IMUData* magnetometerValues)
+{
+    if (nullptr != magnetometerValues)
+    {
+        magnetometerValues->valueX = m_magnetometerValues.valueX;
+        magnetometerValues->valueY = m_magnetometerValues.valueY;
+        magnetometerValues->valueZ = m_magnetometerValues.valueZ;
+    }
+}
+
+void IMU::calibrate()
+{
+    // TODO: implement TD067
+}
+
+int16_t IMU::convertFromDoubleToInt16(double originalValue)
+{
+    int16_t       convertedValue;
+    const int16_t minimumValue = std::numeric_limits<int16_t>::min();
+    const int16_t maximumValue = std::numeric_limits<int16_t>::max();
+
+    if (originalValue <= minimumValue)
+    {
+        convertedValue = minimumValue;
```

Comment:
No scaling required?

Reply:
The Look-Up-Table of the Sensors in the Zumo32U4.proto file is adapted to match the response curve of the physical sensors. This function is only for the conversion from the double value (that the webots library delivers) to an integer value and to prevent a possible overflow.